src/ansys_sphinx_theme/navbar_dropdown.py

```diff
@@ -0,0 +1,135 @@
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+"""Navigation Dropdown for navigation bar."""
+
+import copy
+from functools import lru_cache
+import pathlib
+from typing import Dict, List, Union
+
+import bs4
+from docutils import nodes
+import sphinx
+from sphinx.util import logging
+from sphinx.util.nodes import make_refnode
+import yaml
+
+logger = logging.getLogger(__name__)
+
+
+def load_navbar_configuration(app: sphinx.application.Sphinx) -> None:
+    """Load the navbar configuration from a YAML file for the Sphinx app."""
+    if not (
+        "navigation_dropdown" in app.config.html_theme_options
+        and "layout_file" in app.config.html_theme_options["navigation_dropdown"]
+    ):
+        return
+
+    layout_file = app.config.html_theme_options["navigation_dropdown"].get("layout_file", None)
+    try:
+        with pathlib.Path.open(app.srcdir / layout_file, encoding="utf-8") as config_file:
+            app.config.navbar_contents = yaml.safe_load(config_file)
+    except FileNotFoundError:
+        raise FileNotFoundError(f"Could not find {layout_file}.")
+    except yaml.YAMLError as exc:
+        raise ValueError(f"Error parsing '{layout_file}': {exc}")
+
+
+NavEntry = Dict[str, Union[str, List["NavEntry"]]]
+"""Type alias for a navigation entry in the navbar configuration.
+
+Each entry can have a 'file' or 'link' key, and optionally 'title',
+'caption', and 'sections' keys. The 'sections' key contains a list of
+sub-entries, allowing for nested navigation structures.
+"""
+
+
+def update_template_context(app, pagename, templatename, context, doctree):
+    """Inject custom variables and utilities into the Sphinx template context."""
+
+    @lru_cache(maxsize=None)
+    def render_navbar_links_html() -> bs4.BeautifulSoup:
+        """Render external header links as HTML for the navbar."""
+        if not hasattr(app.config, "navbar_contents"):
+            raise ValueError("Navbar configuration is missing. Please specify a navbar YAML file.")
+        node = nodes.container(classes=["navbar-content"])
+        node.append(build_navbar_nodes(app.config.navbar_contents))
+        header_soup = bs4.BeautifulSoup(app.builder.render_partial(node)["fragment"], "html.parser")
+        return add_navbar_chevrons(header_soup)
+
+    def build_navbar_nodes(obj: List[NavEntry], is_top_level: bool = True) -> nodes.Node:
```

Comment:
```suggestion
    def build_navbar_nodes(obj: list[NavEntry], is_top_level: bool = True) -> nodes.Node:
```
Since we support python 3.10+, annotation can be done with built-in types and the use of `Union` can also be avoided.